### PR TITLE
Respect target-cpu in RUSTFLAGS

### PR DESCRIPTION
**Pull request by breezewish**
Partially resolve https://github.com/rust-cross/cargo-zigbuild/issues/241

1. in zigbuild command, target-cpu in the RUSTFLAGS (or local Cargo config files) is parsed and appended to the compiler wrapper.

   For example, `RUSTFLAGS="-Ctarget-cpu=x86-64-v3"` will cause compiler wrapper to be `cargo-zigbuild zig cc -- -g -mcpu=x86_64_v3 ...`

2. -Ctarget-features is currently not respected in this PR. However it is parsed well.

3. Add a test using https://github.com/breezewish/x86-instruction-set-analyzer to check whether desired instruction set presents in the compiled binary. This helps us make sure that RUSTFLAGS are correctly passed.

   Note: As we check instruction sets for the entire binary, it is possible that the instruction set (for example AVX2) comes from the Rust part, not the C++ part. I tested manually that, when `RUSTFLAGS="-C target_cpu=x86-64-v4"` is set, the Rust part will not produce AVX2 instructions for `tests/target-cpu`. If we discovered AVX2 instruction, it must comes from the C++ part.

   Here is the instruction check result with `RUSTFLAGS="-C target_cpu=x86-64-v4"` without using this PR:

   ```
   Format: Elf Little-endian 64-bit
   Kind: Dynamic
   Architecture: X86_64
   
   Instruction set usage (total 144541):
   X64: 93623 (64.77%)
   INTEL386: 31545 (21.82%)
   INTEL8086: 11774 (8.15%)
   SSE: 2660 (1.84%)
   SSE2: 1547 (1.07%)
   CMOV: 1395 (0.97%)
   MULTIBYTENOP: 1280 (0.89%)
   BMI2: 528 (0.37%)
   INTEL286: 46 (0.03%)
   INTEL186: 31 (0.02%)
   BMI1: 31 (0.02%)
   LZCNT: 30 (0.02%)
   INTEL486: 28 (0.02%)
   AVX: 17 (0.01%)
   PAUSE: 4 (0.00%)
   CPUID: 2 (0.00%)
   ```

### CODE REVIEW

**Comment by breezewish** [comment]
Label `CI-no-fail-fast` may be needed in order to verify the newly added test case.

**Comment by breezewish** [comment]
CI is now passed, except for zig-master branch :)

File: .github/workflows/CI.yml [hunk 1]
```diff
@@ -82,6 +82,38 @@ jobs:
           cargo run zigbuild --manifest-path tests/bindgen-exhaustive/Cargo.toml --target aarch64-unknown-linux-gnu
           cargo run zigbuild --manifest-path tests/bindgen-exhaustive/Cargo.toml --target x86_64-pc-windows-gnu
           cargo run zigbuild --manifest-path tests/bindgen-exhaustive/Cargo.toml --target aarch64-apple-darwin
+      - name: Test RUSTFLAGS
+        shell: bash
+        run: |
+          set -e
+          cargo install x86-instruction-set-analyzer --version 0.1.0
```
Comment (messense) [inline .github/workflows/CI.yml:89]: Could you release binaries for this project so we don't need to compile it every time on GitHub Actions?

**Comment by messense** [comment]
> Oh, looks like the newly introduced dependency rustflags required rustc 1.74. Ref [dtolnay/rustflags#5](https://github.com/dtolnay/rustflags/pull/5)

Not a big issue, I'll cut a new release from main before merging this.

File: .github/workflows/CI.yml (hunk 1 above)
Comment (breezewish) [reply .github/workflows/CI.yml]: Thanks! Really good project. Now x86-instruction-set-analyzer is binary installed in CI.

File: src/zig.rs [hunk 2]
```diff
@@ -963,6 +963,43 @@ pub struct ZigWrapper {
...
+        for flag in f {
+            if let rustflags::Flag::Codegen { opt, value } = flag {
+                let key = opt.replace('-', "_");
+                match key.as_str() {
+                    "target_cpu" => {
+                        if let Some(value) = value {
+                            parsed.target_cpu = value;
+                        }
+                    }
+                    "target_feature" => {
```
Comment (messense) [inline src/zig.rs:986]: No point in replacing `-` with `_` here?

```suggestion
                match opt.as_str() {
                    "target-cpu" => {
                        if let Some(value) = value {
                            parsed.target_cpu = value;
                        }
                    }
                    "target-feature" => {
```

**Review by messense — APPROVED** [review]
Mostly LGTM, just a minor nit.

File: src/zig.rs (hunk 2 above)
Comment (breezewish) [reply src/zig.rs:986]: Both -Ctarget_cpu and -Ctarget-cpu are recognizable by rustc. Also see https://github.com/rust-lang/rust/blob/e7da0fa62fc58f9d665a67bf2e7a6d6154d8d74c/compiler/rustc_session/src/options.rs#L340

Comment (messense) [reply src/zig.rs:986]: OK, thanks for the explanation.

**Pull request merged**